libbeat/common/diagnostics/diagnostics.go

```diff
@@ -0,0 +1,48 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package diagnostics
+
+// DiagnosticRunner is a sub-type of the runner interface found as cfgfile.Runner
+// Runners can implement this if they want their client filesets/metricsets to expose the underlying DiagnosticSet interface
+type DiagnosticRunner interface {
+	ModuleDiagnostics() []DiagnosticSetup
+}
+
+// DiagnosticSet is an interface that a metricset/fileset can implement to provide additional Diagnostic data.
+// A metricset can provide any number of diagnostic responses when requested.
+type DiagnosticSet interface {
+	// DiagnosticSetup returns metadata and a callback halder.
+	// note that this can be called any time after a metricset has started, so implementors should not assume
+	// the state of a metricset/fileset when this method is called.
+	DiagnosticSetup() []DiagnosticSetup
+}
+
+// DiagnosticSetup contains the data needed to register a callback.
+type DiagnosticSetup struct {
+	// The name of this diagnostics data result.
+	Name string
+	// A brief description of the file.
+	Description string
+	// The filename that the requester should save the body as. This value must be unique for all other diagnostics in the metricset/fileset
+	Filename string
+	// MIME/ContentType. See https://www.iana.org/assignments/media-types/media-types.xhtml
+	ContentType string
+	//Callback is called when diagnostic data is actually requested by central management.
+	// Callback does not return an error, and if one occours, it should be written out as the result.
+	Callback func() []byte
```

Comment:
My big-picture wish is that we could skip the whole `Callback` indirection and just return a list of filenames + data directly from the main `Diagnostics()` call -- that is, have the call return the actual current diagnostics instead of a list of possible future diagnostics to add hooks for. I don't know of a reason to register every possible file in advance when we could just return a list of the relevant ones when `diagnostics` is actually invoked, since the registration seems to only be used internally to the client, not by Agent itself. But it would require new API in `elastic-agent-client`, so this approach is probably necessary until we address that...

Reply:
Yup, agreed. Would be simpler to just not do the callback thing, but I didn't want to invent some kind of overly-clever wrapper around `RegisterDiagnosticHook`.